### ISSUE TITLE
authentication

### ISSUE THREAD

**Issue by yieme** [js-data/js-data-firebase#2]
Like the objectives of the js-data project. Looking through the docs and code and I'm not seeing a way to authenticate the js-data-firebase, ie: ref.authWithCustomToken(), or initialize js-data-firebase with an authenticated firebase reference instead just the firebase base url. Thoughts?

**Comment by jmdobry** [comment]
Currently js-data-firebase creates lots of refs and throws them away. I would have to refactor to a single ref which would then be exposed so you can do auth.

**Comment by jmdobry** [comment]
I honestly haven't used Firebase authentication much. Once you log a user in does it persist across `ref` instances?

**Comment by kentcdodds** [comment]
I believe it does.

**Comment by jmdobry** [comment]
If that's the case, doesn't the user just need to create their own `ref` object to do authentication (as long as it has the same domain as the `basePath` given to the `DSFirebaseAdapter` constructor)?

**Comment by kentcdodds** [comment]
I'm pretty sure that I've done that before...

Yes, I'm doing it in issuetemplate.com. I create one ref [here](https://github.com/kentcdodds/issue-template/blob/86856f8fc34f728fd7f5a107c64a045078ea78c2/app/services/LoginService.js#L2) to authenticate and then another ref [here](https://github.com/kentcdodds/issue-template/blob/86856f8fc34f728fd7f5a107c64a045078ea78c2/app/services/TemplateService.js#L2) to get stuff.

**Comment by jmdobry** [comment]
@yieme Can you try something like the following and see if it works?

```js
// Create store and adapter
var store = new JSData.DS();
var adapter = new DSFirebaseAdapter({
  basePath: 'https://my-app.firebase.io'
});
store.registerAdapter('fb', adapter, { default: true });

var ref = new Firebase('https://my-app.firebase.io');
ref. authWithCustomToken();
```

I'm thinking that by authenticating with your own `ref` it will also authenticate for the adapter because you're using the same `"https://my-app.firebase.io"`.

**Comment by yieme** [comment]
Good thought. Didn't seem to work. Sample code:

```js
var adapter = new DSFirebaseAdapter({ basePath: FIREBASE_URL })
var store   = new JSData.DS()
store.registerAdapter('fb', adapter, { default: true })
var ref     = new Firebase(FIREBASE_URL)
function AuthDone(err, data) {
  console.log('Authenticated', err, data)
    ref.child('person').push({ source: 'Firebase' }, function () { console.log('push ok') })
  var People  = store.defineResource('person')
  var people  = People.create({ source: 'JSData' }).then(function(doc) {
    console.log('doc:', doc)
  })
}
ref.authWithCustomToken(AUTH_TOKEN, AuthDone)
```

Output:

```
Authenticated null { auth: null,
  expires: null,
  token: '',
  uid: null,
  provider: 'custom' }
push ok
FIREBASE WARNING: set at /person/-JewmzxPgIwCgb0cQ7QD failed: permission_denied 
Possibly unhandled Error: PERMISSION_DENIED: Permission denied
    at Error (<anonymous>)
```

Resulting firebase data for people:

```js
{
  "-JewlzN7FtMOFtpS6n01" : {
    "source" : "Firebase"
  }
}
```

**Comment by jmdobry** [comment]
I'll have to take a closer look at this then.

**Comment by yieme** [comment]
FYI adding ```this.defaults.firebaseRef || ``` to the beginning of index.js line 47 to the following works:

```js
  return this.defaults.firebaseRef || new Firebase(makePath(options.basePath || this.defaults.basePath || resourceConfig.basePath, options.endpoint || resourceConfig.endpoint));
```

```js
var adapter = new DSFirebaseAdapter({ firebaseRef: authenticatedFirebaseRef })
```

**Comment by jmdobry** [comment]
@yieme I just released 1.0.0-beta.1. Now, an instance DSFirebaseAdapter exposes the `ref` object on which you can call `authWithCustomToken`.

```js
var fb = new DSFirebaseAdapter({
  basePath: 'https://my-firebase-url.firebaseio.com'
});

fb.ref.authWithCustomToken(...);
```

**Comment by Bartozzz** [comment]
For anybody struggling the same problem, I've created a `React/js-data` starter which includes basic Firebase authentication. Hope it helps!

Demo: https://fir-starter-4d822.firebaseapp.com/#/
Repo: https://github.com/project-inra/react-js-data-starter
